File: tdms/tests/system/data/input_generation/matlab/multi_layer.m
function [S] = multi_layer(zvec, nvec, lambda0, theta0, sp)
    %% Compute the characteristic matrix of a multilayer structure.
    % zvec      : [    z1, z2, ..., zN].
    % nvec      : [n0, n1, n2, ..., nN].
    % lambda0   : Wavelength of light in air.
    % theta0    : Angle of incidence of the plane wave, in radians.
    % sp        : Either 'TE' or 'TM'. 'TE' is taken to mean that the electric vector is perpendicular to the plane in which the wave propagates. 'TM' is when the electric field vector is in, or parallel to, the plane in which the wave propagates.
    %
    % S         : Characteristic matrix of the multilayer structure.

%% Check dimensionality of inputs
if numel(nvec)-numel(zvec) ~= 1
    error('nvec should have one more element that zvec');
end

%% Compute constant values
theta_vec=asin(nvec(1)*sin(theta0)./nvec);
st=sin(theta_vec);
ct=cos(theta_vec);

d=diff(zvec);
if numel(nvec)>2
    beta=ct(2:(end-1)).*nvec(2:(end-1)).*d*2*pi/lambda0;
end

if strcmp(sp,'TE')
    t_fresnel = 2*nvec(1:(end-1)).*ct(1:(end-1))./(nvec(1:(end-1)).*ct(1:(end-1))+nvec(2:end).*ct(2:end));
    r_fresnel = (nvec(1:(end-1)).*ct(1:(end-1))-nvec(2:end).*ct(2:end))./(nvec(1:(end-1)).*ct(1:(end-1))+nvec(2:end).*ct(2:end));
else
    t_fresnel = 2*nvec(1:(end-1)).*ct(1:(end-1))./(nvec(2:end).*ct(1:(end-1))+nvec(1:(end-1)).*ct(2:end));
    r_fresnel = (nvec(2:end).*ct(1:(end-1))-nvec(1:(end-1)).*ct(2:end))./(nvec(2:end).*ct(1:(end-1))+nvec(1:(end-1)).*ct(2:end));
end

%% Setup scattering matrix
S = 1/t_fresnel(1)*[1 r_fresnel(1);r_fresnel(1) 1];
for il=2:numel(zvec)
    P = [exp(-sqrt(-1)*beta(il-1)) 0;0 exp(sqrt(-1)*beta(il-1))];
    L = 1/t_fresnel(il)*[1 r_fresnel(il);r_fresnel(il) 1];
    S=S*P*L;
end
end

File: tdms/tests/system/py_input_regen/misc_functions/multi_layer.py
from typing import Literal

import numpy as np


def multi_layer(
    zvec: np.ndarray,
    nvec: np.ndarray,
    wavelength_in_air: float,
    theta0,
    polarisation: Literal["TE", "TM"] = "TE",
) -> np.ndarray:
    """
    Compute the characteristic matrix of a multilayer structure.

    The input array nvec should have one more element than the zvec input.

    :param zvec: Vector defining the z-locations of the layers of the (unrotated) stratified medium.
    :param nvec: Refractive indices of the layers.
    :param wavelength_in_air: Wavelength of light in air.
    :param theta0: Angle of incidence of the plane wave, in radians.
    :param polarisation: The polarisation of the source. 'TE' means that the electric vector is perpendicular to the plane in which the wave propagates. 'TM' is used when the electric field is parallel to, or in, the plane in which the wave propagates.
    :returns: The characteristic matrix of the multiplayer structure.
    """
    # Check dimensionality of input
    if nvec.size != (zvec.size + 1):
        raise RuntimeError(
            f"nvec should have one more element than zvec, but they have {nvec.size} and {zvec.size} elements respectively."
        )

    # Constant values
    theta_vec = np.arcsin(nvec[0] * np.sin(theta0) / nvec)
    cos_theta = np.cos(theta_vec)
    # Speedup to save recomputing multiple times
    intermediate_product = nvec * cos_theta

    if nvec.size > 2:
        beta = (
            intermediate_product[1:-1] * np.diff(zvec) * 2.0 * np.pi / wavelength_in_air
        )

    if polarisation == "TE":
        denominator = intermediate_product[:-1] + intermediate_product[1:]
        t_fresnel = 2.0 * intermediate_product[:-1] / denominator
        r_fresnel = -np.diff(intermediate_product) / denominator
    else:  # polarisation == 'TM'
        denominator = nvec[1:] * cos_theta[:-1] + nvec[:-1] * cos_theta[1:]
        t_fresnel = 2.0 * intermediate_product[:-1] / denominator
        r_fresnel = (
            nvec[1:] * cos_theta[:-1] - nvec[:-1] * cos_theta[1:]
        ) / denominator

    # Define characteristic/scattering matrix
    S = 1 / t_fresnel[0] * np.array([[1.0, r_fresnel[0]], [r_fresnel[0], 1.0]])
    for i in range(1, nvec.size - 1):
        # Note that beta is not defined if nvec has 2 or fewer elements.
        # However in this case, this loop doesn't execute.
        P = np.diag(np.exp([-1.0j * beta[i - 1], 1.0j * beta[i - 1]]))
        L = 1 / t_fresnel[i] * np.array([[1, r_fresnel[i]], [r_fresnel[i], 1]])
        # S = S * P * L
        S = np.matmul(np.matmul(S, P), L)

    return S
